// artik-sdk/src/platform/eagleye530.js

/* GPIO definition */
exports.ARTIK_EAGLEYE530_GPIO_GPIO0 = 128;
exports.ARTIK_EAGLEYE530_GPIO_GPIO1 = 129;
exports.ARTIK_EAGLEYE530_GPIO_GPIO2 = 130;
exports.ARTIK_EAGLEYE530_GPIO_GPIO3 = 46;
exports.ARTIK_EAGLEYE530_GPIO_GPIO4 = 14;
exports.ARTIK_EAGLEYE530_GPIO_GPIO6 = 25;
exports.ARTIK_EAGLEYE530_GPIO_GPIO7 = 0;
exports.ARTIK_EAGLEYE530_GPIO_GPIO8 = 26;
exports.ARTIK_EAGLEYE530_GPIO_GPIO9 = 27;
exports.ARTIK_EAGLEYE530_GPIO_AGPIO0 = 161;

/* LED definition */
exports.ARTIK_EAGLEYE530_GPIO_LED1 = 159;
exports.ARTIK_EAGLEYE530_GPIO_LED2 = 43;

/* SWITCH definition */
exports.ARTIK_EAGLEYE530_GPIO_SW1 = 50;
exports.ARTIK_EAGLEYE530_GPIO_SW2 = 54;

/* SERIAL definition */
exports.ARTIK_EAGLEYE530_SERIAL = {
	'UART': {
		'UART4': 4
	}
};

/* PWM definition */
exports.ARTIK_EAGLEYE530_PWM = {
	'PWMIO': {
		'XPWMIO0': 0,
		'XPWMIO2': 2,
	},
	'POLR': {
		'NORMAL' : 0,
		'INVERT' : 1
	}
};

/* SPI definition */
exports.ARTIK_EAGLEYE530_SPI = {
	'BUS': {
		'BUS0': 2
	},
	'CS': {
		'CS0' : 0
	},
	'MODE': {
		'MODE0' : 0,
		'MODE1' : 1,
		'MODE2' : 2,
		'MODE3' : 3,
	},
	'BITS': {
		'BITS8': 8
	},
};
